netplan/libnetplan.py

```diff
@@ -22,12 +22,128 @@
 import ctypes.util
 from ctypes import c_char_p, c_void_p, c_int, c_uint, c_size_t, c_ssize_t
 from typing import List, Union, IO
+from enum import Enum
+import re
 
 
-class LibNetplanException(Exception):
+SCHEMA_ERROR_MSG_REGEX = (
+    r'(?P<file_path>.*):(?P<error_line>\d+):(?P<error_col>\d+): (?P<message>.*)'
+)
+
+
+class NETPLAN_ERROR_DOMAINS(Enum):
+    # Errors and error domains
+    NETPLAN_PARSER_ERROR = 1
+    NETPLAN_VALIDATION_ERROR = 2
+    NETPLAN_FILE_ERROR = 3
+    NETPLAN_BACKEND_ERROR = 4
+
+    @classmethod
+    def is_parser_error(cls, domain):
+        return cls.NETPLAN_PARSER_ERROR.value == domain
+
+    @classmethod
+    def is_validation_error(cls, domain):
+        return cls.NETPLAN_VALIDATION_ERROR.value == domain
+
+    @classmethod
+    def is_file_error(cls, domain):
+        return cls.NETPLAN_FILE_ERROR.value == domain
+
+
+class NETPLAN_PARSER_ERRORS(Enum):
+    # Errors for domain NETPLAN_PARSER_ERROR
+    NETPLAN_ERROR_INVALID_YAML = 0
+
+
+class NETPLAN_VALIDATION_ERRORS(Enum):
+    # Errors for domain NETPLAN_VALIDATION_ERROR
+    NETPLAN_ERROR_CONFIG_VALIDATION = 0
+    NETPLAN_ERROR_CONFIG_GENERIC = 1
+
+    @classmethod
+    def is_config_validation_error(cls, error):
+        return error == cls.NETPLAN_ERROR_CONFIG_VALIDATION.value
+
+
+class NETPLAN_FILE_ERRORS(Enum):
+    # Errors for domain NETPLAN_FILE_ERROR
+    NETPLAN_ERROR_FILE_PERMISSION_DENIED = 0
+    NETPLAN_ERROR_FILE_CREATE = 1
+    NETPLAN_ERROR_FILE_OPEN = 2
+    NETPLAN_ERROR_FILE = 3
+
+
+class NETPLAN_BACKEND_ERRORS(Enum):
+    # Errors for domain NETPLAN_BACKEND_ERROR
+    NETPLAN_ERROR_UNSUPPORTED = 0
+    NETPLAN_ERROR_VALIDATION = 1
+
+
+class NetplanException(Exception):
+    def __init__(self, domain=None, error=None, filename=None, line=None, column=None, message=None):
+        self.domain = domain
+        self.error = error
+        self.filename = filename
+        self.line = line
+        self.column = column
+        self.message = message
+
+
+class LibNetplanException(NetplanException):
     pass
 
 
+class LibNetplanFileException(NetplanException):
```

Comment:
Agree. But I decided to keep the "Lib" for now just because I also decided to keep the LibNetplanException.

If we agree that it's OK to drop LibNetplanException in favor of NetplanException I will also drop the "Lib" prefix.